langchain/src/schema/load.ts

```diff
@@ -0,0 +1,105 @@
+export interface SerializedFields {
+  [key: string]: unknown;
+}
+
+export interface BaseSerialized<T extends string> {
+  lc: number;
+  type: T;
+  id: string[];
+}
+
+export interface SerializedConstructor extends BaseSerialized<"constructor"> {
+  arguments: unknown[];
+  fields?: SerializedFields;
+}
+
+export interface SerializedFunction extends BaseSerialized<"function"> {
+  arguments: unknown[];
+}
+
+export interface SerializedSecret extends BaseSerialized<"secret"> {}
+
+export interface SerializedNotImplemented
+  extends BaseSerialized<"not_implemented"> {}
+
+export type Serialized =
+  | SerializedConstructor
+  | SerializedFunction
+  | SerializedSecret
+  | SerializedNotImplemented;
+
+function shallowCopy<T extends object>(obj: T): T {
+  return Array.isArray(obj) ? ([...obj] as T) : ({ ...obj } as T);
+}
+
+function replaceSecrets(
+  root: unknown[],
+  secretsMap: { [key: string]: string }
+): unknown[] {
+  const result = shallowCopy(root);
+  for (const [path, secretId] of Object.entries(secretsMap)) {
+    const [last, ...partsReverse] = path.split(".").reverse();
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    let current: any = result;
+    for (const part of partsReverse.reverse()) {
+      current[part] = shallowCopy(current[part]);
+      current = current[part];
+    }
+    if (current[last] !== undefined) {
+      current[last] = {
+        lc: 1,
+        type: "secret",
+        id: [secretId],
+      };
+    }
+  }
+  return result;
+}
+
+export abstract class Serializable {
+  abstract lc_namespace: string[];
+
+  lc_arguments: unknown[];
+
+  get lc_secrets(): { [key: string]: string } | undefined {
```

Comment:
What's the purpose behind the `0.SECRET_NAME`?

Reply:
It's to prevent those values from being included in the serialised object, eg api keys

Reply:
Because it's in the first argument to the constructor, thus 0. If you hunt around all the files you'll see some constructors (sadly) take more than one arg